video/out/opengl/user_shaders.c

```diff
@@ -287,3 +277,163 @@ bool parse_user_shader_pass(struct mp_log *log, struct bstr *body,
+        }
+
+        if (bstr_eatstart0(&line, "FILTER")) {
+            line = bstr_strip(line);
+            if (bstr_equals0(line, "LINEAR")) {
+                out->gl_filter = GL_LINEAR;
+            } else if (bstr_equals0(line, "NEAREST")) {
+                out->gl_filter = GL_NEAREST;
+            } else {
+                mp_err(log, "Unrecognized FILTER: '%.*s'!\n", BSTR_P(line));
+                return false;
+            }
+            continue;
+        }
+
+        if (bstr_eatstart0(&line, "BORDER")) {
+            line = bstr_strip(line);
+            if (bstr_equals0(line, "CLAMP")) {
+                out->gl_border = GL_CLAMP_TO_EDGE;
+            } else if (bstr_equals0(line, "REPEAT")) {
+                out->gl_border = GL_REPEAT;
+            } else if (bstr_equals0(line, "MIRROR")) {
+                out->gl_border = GL_MIRRORED_REPEAT;
+            } else {
+                mp_err(log, "Unrecognized BORDER: '%.*s'!\n", BSTR_P(line));
+                return false;
+            }
+            continue;
+        }
+
+        mp_err(log, "Unrecognized command '%.*s'!\n", BSTR_P(line));
+        return false;
+    }
+
+    // Decode the rest of the section (up to the next //! marker) as raw hex
+    // data for the texture
+    struct bstr hexdata;
+    if (bstr_split_tok(*body, "//!", &hexdata, body)) {
+        // Make sure the magic line is part of the rest
+        body->start -= 3;
+        body->len += 3;
+    }
+
+    struct bstr tex;
+    if (!bstr_decode_hex(NULL, bstr_strip(hexdata), &tex)) {
+        mp_err(log, "Error while parsing TEXTURE body: must be a valid "
+                    "hexadecimal sequence, on a single line!\n");
+        return false;
+    }
+
+    int expected_len = out->w * out->h * out->d * out->components * out->bytes;
+    if (tex.len != expected_len) {
+        mp_err(log, "Shader TEXTURE size mismatch: got %zd bytes, expected %d!\n",
+               tex.len, expected_len);
+        talloc_free(tex.start);
+        return false;
+    }
+
+    out->texdata = tex.start;
+    return true;
+}
+
+void parse_user_shader(struct mp_log *log, struct bstr shader, void *priv,
+                       void (*dohook)(void *p, struct gl_user_shader_hook hook),
+                       void (*dotex)(void *p, struct gl_user_shader_tex tex))
+{
+    if (!dohook || !dotex || !shader.len)
+        return;
+
+    // Skip all garbage (e.g. comments) before the first header
+    int pos = bstr_find(shader, bstr0("//!"));
+    if (pos < 0) {
+        mp_warn(log, "Shader appears to contain no headers!\n");
+        return;
+    }
+    shader = bstr_cut(shader, pos);
+
+    // Loop over the file
+    while (shader.len > 0)
+    {
+        // Peek at the first header to dispatch the right type
+        if (bstr_startswith0(shader, "//!TEXTURE")) {
+            struct gl_user_shader_tex t;
+            if (!parse_tex(log, &shader, &t))
+                return;
+            dotex(priv, t);
+        }
```

Comment:
Good point. I had also wanted to change it to make `dotex` return a bool to break on failure. Added both of these changes to d29208f.